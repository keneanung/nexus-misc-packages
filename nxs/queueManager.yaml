# yaml-language-server: $schema=../node_modules/@keneanung/nexus-package-builder/resources/nexus-schema.json
name: QueueManager
enabled: true
description: A queue manager for Iron Realms Nexus 3
type: group
website: https://keneanung.github.io/nexus-misc-packages
items:
  - type: function
    enabled: true
    name: onLoad
    codeFile: ./queueManager.js
  - type: function
    code: run_function('onLoad', undefined, 'QueueManager')
    name: onInstall
    enabled: true

