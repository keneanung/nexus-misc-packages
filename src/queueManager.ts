interface QueuedItem {
  command: string;
  properties: QueuedItemProperties;
}

interface QueuedItemProperties {
  haveBalance?: boolean;
  haveEq?: boolean;
  haveClassBalance?: boolean;
  haveShipBalance?: boolean;
  haveParalysis?: boolean;
  beBound?: boolean;
  beStanding?: boolean;
  beStunned?: boolean;
}

const itemPropertiesEqual = (one: QueuedItemProperties, other: QueuedItemProperties) => {
  const keys = Object.keys(one);
  if (keys.length !== Object.keys(other).length) {
    return false;
  }
  const castedKeys = Object.keys(one) as (keyof typeof one)[];
  return castedKeys.every((key) => one[key] === other[key]);
};

interface CustomQueueComponent {
  letter: string;
  property: keyof QueuedItemProperties;
}

const customQueueTypeComponents: CustomQueueComponent[] = [
  {
    letter: 'e',
    property: 'haveEq',
  },
  {
    letter: 'b',
    property: 'haveBalance',
  },
  {
    letter: 'c',
    property: 'haveClassBalance',
  },
  {
    letter: 's',
    property: 'haveShipBalance',
  },
  {
    letter: 'p',
    property: 'haveParalysis',
  },
  {
    letter: 'w',
    property: 'beBound',
  },
  {
    letter: 'u',
    property: 'beStanding',
  },
  {
    letter: 't',
    property: 'beStunned',
  },
];

interface Command {
  command: string;
  queue: string;
}

export class QueueManager {
  private queue: QueuedItem[] = [];

  public track(command: string, queue: string) {
    const itemProperties = this.parseQueue(queue);
    this.queue.push({ command, properties: itemProperties });
  }

  private parseQueue(queue: string): QueuedItemProperties {
    if (queue === 'equilibrium') {
      queue = 'e';
    }
    if (queue === 'balance') {
      queue = 'b';
    }
    if (queue === 'class') {
      queue = 'c';
    }
    if (queue === 'paralysis') {
      queue = '!p';
    }
    if (queue === 'unbound') {
      queue = '!w';
    }
    if (queue === 'stun') {
      queue = '!t';
    }
    if (queue === 'free') {
      queue = 'be!p!t';
    }
    if (queue === 'freestand') {
      queue = 'be!p!tu';
    }
    if (queue === 'full') {
      queue = 'be!p!tuc';
    }
    const customProperties: QueuedItemProperties = {};
    for (const queueType of customQueueTypeComponents) {
      const index = queue.indexOf(queueType.letter);
      if (index > -1) {
        const queueTypeValue = index === 0 || queue.at(index - 1) !== '!';
        customProperties[queueType.property] = queueTypeValue;
      }
    }
    return customProperties;
  }

  public getQueue = () => this.queue;

  public sync = (newCommands: Command[]) => {
    this.clear('all');
    for (const { command, queue } of newCommands) {
      this.track(command, queue);
    }
  };

  public clear = (queue: string) => {
    if (queue === 'all') {
      this.queue = [];
      return;
    }
    const queueProps = this.parseQueue(queue);
    for (let i = this.queue.length - 1; i >= 0; i--) {
      if (itemPropertiesEqual(queueProps, this.queue[i].properties)) {
        this.queue.splice(i, 1);
      }
    }
  };

  public trackFirst = (command: string, queue: string) => {
    const itemProperties = this.parseQueue(queue);
    this.queue.unshift({ command, properties: itemProperties });
  };

  public trackAt = (position: number, command: string, queue: string) => {
    const itemProperties = this.parseQueue(queue);
    this.queue.splice(position - 1, 0, { command, properties: itemProperties });
  };

  public trackReplace = (position: number, command: string, queue: string) => {
    const itemProperties = this.parseQueue(queue);
    this.queue.splice(position - 1, 1, { command, properties: itemProperties });
  };

  public trackRemove = (position: number) => {
    this.queue.splice(position - 1, 1);
  };

  public trackAddClear = (command: string, queue: string, full = false) => {
    this.clear(full ? 'all' : queue);
    this.track(command, queue);
  };

  public run = (command: string, queue: string) => {
    const itemProperties = this.parseQueue(queue);
    for (let i = 0; i < this.queue.length; i++) {
      const queuedItem = this.queue[i];
      if (queuedItem.command == command && itemPropertiesEqual(queuedItem.properties, itemProperties)) {
        this.queue.splice(i, 1);
        break;
      }
    }
  };
}
