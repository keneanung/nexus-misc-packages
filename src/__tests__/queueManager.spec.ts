import { QueueManager } from '../queueManager';

test.each([
  ['e'],
  ['b'],
  ['c'],
  ['s'],
  ['p'],
  ['w'],
  ['u'],
  ['t'],
  ['!e'],
  ['!b'],
  ['!c'],
  ['!s'],
  ['!p'],
  ['!w'],
  ['!u'],
  ['!t'],
  ['equilibrium'],
  ['balance'],
  ['class'],
  ['paralysis'],
  ['unbound'],
  ['stun'],
  ['free'],
  ['freestand'],
  ['full'],
  ['frl'],
])('Should translate property %p correctly', (queue: string) => {
  const sut = new QueueManager();

  sut.track('foo', queue);

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have an empty queue after clear("all")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('all');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have remove items in class queue after clear("class")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('class');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should have remove items in class queue after clear("c")', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.clear('c');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should put prepended items first in queue', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackFirst('bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should put items at given position in queue with trackAt', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackAt(3, 'bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should replace items at given position in queue with trackReplace', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackReplace(2, 'bamboozle', 'freestand');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove items at given position in queue with trackRemove', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'class');
  sut.track('baz', 's');

  sut.trackRemove(2);

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('foo', 'full');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue with different casing', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('FOO', 'full');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Should remove run command from queue (not first item)', () => {
  const sut = new QueueManager();
  sut.track('foo', 'full');
  sut.track('bar', 'full');
  sut.track('baz', 'full');
  sut.track('foo', 'free');
  sut.track('boom', 'free');

  sut.run('foo', 'free');

  expect(sut.getQueue()).toMatchSnapshot();
});

test('Workaround for in-game bug #17807', () => {
  const sut = new QueueManager();
  sut.track('sit', 'full');
  sut.track('stand', 'full');
  sut.track('sit', 'eu');

  sut.run('sit', 'eu');
  sut.run('stand', 'eu');
  sut.run('sit', 'eu');

  expect(sut.getQueue()).toMatchSnapshot();
});
