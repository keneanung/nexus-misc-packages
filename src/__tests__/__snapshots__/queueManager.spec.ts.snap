// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have an empty queue after clear("all") 1`] = `[]`;

exports[`Should have remove items in class queue after clear("c") 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should have remove items in class queue after clear("class") 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should put items at given position in queue with trackAt 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "bar",
    "properties": {
      "haveClassBalance": true,
    },
  },
  {
    "command": "bamboozle",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should put prepended items first in queue 1`] = `
[
  {
    "command": "bamboozle",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "bar",
    "properties": {
      "haveClassBalance": true,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should replace entire queue on sync() 1`] = `
[
  {
    "command": "bar",
    "properties": {
      "haveClassBalance": true,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should replace items at given position in queue with trackReplace 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "bamboozle",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
  {
    "command": "baz",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should translate property "!b" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveBalance": false,
    },
  },
]
`;

exports[`Should translate property "!c" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveClassBalance": false,
    },
  },
]
`;

exports[`Should translate property "!e" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveEq": false,
    },
  },
]
`;

exports[`Should translate property "!p" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should translate property "!s" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveShipBalance": false,
    },
  },
]
`;

exports[`Should translate property "!t" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStunned": false,
    },
  },
]
`;

exports[`Should translate property "!u" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": false,
    },
  },
]
`;

exports[`Should translate property "!w" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beBound": false,
    },
  },
]
`;

exports[`Should translate property "b" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveBalance": true,
    },
  },
]
`;

exports[`Should translate property "balance" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveBalance": true,
    },
  },
]
`;

exports[`Should translate property "c" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveClassBalance": true,
    },
  },
]
`;

exports[`Should translate property "class" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveClassBalance": true,
    },
  },
]
`;

exports[`Should translate property "e" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveEq": true,
    },
  },
]
`;

exports[`Should translate property "equilibrium" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveEq": true,
    },
  },
]
`;

exports[`Should translate property "free" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should translate property "freestand" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should translate property "frl" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {},
  },
]
`;

exports[`Should translate property "full" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
      "beStunned": false,
      "haveBalance": true,
      "haveClassBalance": true,
      "haveEq": true,
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should translate property "p" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveParalysis": true,
    },
  },
]
`;

exports[`Should translate property "paralysis" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveParalysis": false,
    },
  },
]
`;

exports[`Should translate property "s" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "haveShipBalance": true,
    },
  },
]
`;

exports[`Should translate property "stun" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStunned": false,
    },
  },
]
`;

exports[`Should translate property "t" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStunned": true,
    },
  },
]
`;

exports[`Should translate property "u" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beStanding": true,
    },
  },
]
`;

exports[`Should translate property "unbound" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beBound": false,
    },
  },
]
`;

exports[`Should translate property "w" correctly 1`] = `
[
  {
    "command": "foo",
    "properties": {
      "beBound": true,
    },
  },
]
`;
